### PR TITLE
Fix Array#[] with ArithmeticSequence with negative steps

### PR DESCRIPTION
**Pull request by jeremyevans**
Previously, Array#[] when called with an ArithmeticSequence
with a negative step did not handle all cases correctly,
especially cases involving infinite ranges, inverted ranges,
and/or exclusive ends.

Fixes [Bug #18247]

### CODE REVIEW

File: spec/ruby/core/array/shared/slice.rb [hunk 1]
```diff
@@ -784,6 +784,102 @@ def to.to_int() -2 end
...
+        @array.send(@method, (-2..-5).step(-4)).should == [4]
+        @array.send(@method, (-2..-5).step(-5)).should == [4]
+
+        @array.send(@method, (-2...-5).step(-1)).should == [4, 3, 2]
+        @array.send(@method, (-2...-5).step(-2)).should == [4, 2]
+        @array.send(@method, (-2...-5).step(-3)).should == [4]
+        @array.send(@method, (-2...-5).step(-4)).should == [4]
+      end
+    end
+  end
```
Comment (mrkn) [inline spec/ruby/core/array/shared/slice.rb:881]: @jeremyevans Could you please put just all the edge cases of them (e.g. the case of `(2..).step(-4)`) in `test_slice` in `test/ruby/test_array.rb`, too?

File: enumerator.c [hunk 2]
```diff
@@ -3802,6 +3802,13 @@ rb_arithmetic_sequence_beg_len_step(VALUE obj, long *begp, long *lenp, long *ste
     *stepp = step;
 
     if (step < 0) {
+        if (aseq.exclude_end && !NIL_P(aseq.end)) {
+            /* Handle exclusion before range reversal */
+            aseq.end = LONG2NUM(NUM2LONG(aseq.end) + 1);
+
+            /* Don't exclude the previous beginning */
+            aseq.exclude_end = 0;
+        }
```
Comment (mrkn) [inline enumerator.c:3811]: We need to write tests of `rb_arithmetic_sequence_beg_len_step` at least to check the behavior of this case because this is a public C-API.
Could you please write `ext/-test-/arith_seq/beg_len_step` and `test/-ext-/arith_seq/test_beg_len_step.rb`?

**Review by mrkn — CHANGES_REQUESTED** [review]
@jeremyevans The implementation and the examples in rubyspec look good to me.
Could you please add some tests I mentioned in the comments I left?

File: spec/ruby/core/array/shared/slice.rb (hunk 1 above)
Comment (jeremyevans) [reply spec/ruby/core/array/shared/slice.rb:881]: I translated the spec code into test code and added it to test_array.

File: enumerator.c (hunk 2 above)
Comment (jeremyevans) [reply enumerator.c:3811]: I added tests for rb_arithmetic_sequence_beg_len_step as requested.

File: ext/-test-/arith_seq/beg_len_step/beg_len_step.c [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+static VALUE
+arith_seq_s_beg_len_step(VALUE mod, VALUE obj, VALUE len, VALUE err)
+{
+  VALUE r;
+  long begp, lenp, stepp;
+
+  r = rb_arithmetic_sequence_beg_len_step(obj, &begp, &lenp, &stepp, NUM2LONG(len), NUM2INT(err));
+
+  return rb_ary_new_from_args(4, r, LONG2NUM(begp), LONG2NUM(lenp), LONG2NUM(stepp));
```
Comment (mrkn) [inline ext/-test-/arith_seq/beg_len_step/beg_len_step.c:11]: I want to remove the suffix `p` from `begp`, `lenp`, and `stepp` because they are not pointers.

```suggestion
  long beg, len, step;

  r = rb_arithmetic_sequence_beg_len_step(obj, &beg, &len, &step, NUM2LONG(len), NUM2INT(err));

  return rb_ary_new_from_args(4, r, LONG2NUM(beg), LONG2NUM(len), LONG2NUM(step));
```

File: ext/-test-/arith_seq/beg_len_step/beg_len_step.c [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+
+static VALUE
+arith_seq_s_beg_len_step(VALUE mod, VALUE obj, VALUE len, VALUE err)
+{
+  VALUE r;
+  long beg, len, step;
+
+  r = rb_arithmetic_sequence_beg_len_step(obj, &beg, &len, &step, NUM2LONG(len), NUM2INT(err));
+
+  return rb_ary_new_from_args(4, r, LONG2NUM(beg), LONG2NUM(len), LONG2NUM(step));
```
Comment (mrkn) [inline ext/-test-/arith_seq/beg_len_step/beg_len_step.c:11]:
```suggestion
  long beg, len2, step;

  r = rb_arithmetic_sequence_beg_len_step(obj, &beg, &len2, &step, NUM2LONG(len), NUM2INT(err));

  return rb_ary_new_from_args(4, r, LONG2NUM(beg), LONG2NUM(len2), LONG2NUM(step));
```

**Review by mrkn — APPROVED** [review]
